### DISCUSSION TITLE
How to maintain/manage declarative podman containers with docker-compose api/yaml and systemd

### DISCUSSION THREAD

**Discussion by andrew-kennedy** [containers/podman#11711] [Q&A]
In issue #11370 I asked the following question:

Is there no official way to do declarative podman image configuration and then systemd unit generation? With docker I could create a set of compose files that I could update and add to as needed for all the services on a system. With podman, I figured I could leverage docker-compose API support to match this using the systemd unit file generation, but it seems this might not be the case? 

I would like to avoid maintaining a set of bash podman create scripts as my set of "declarative" podman images.

**Comment by vrothberg** [comment]
> Is there no official way to do declarative podman image configuration and then systemd unit generation? 

Can you elaborate on that?  I cannot brain parse what that means in practice.

For the limitations of `generate systemd`, see the reply in the linked issue:
https://github.com/containers/podman/issues/11370#issuecomment-909195226

**Reply by andrew-kennedy** [reply]
Sure! What I mean is that I'd like some way to use a configuration language like docker-compose yaml to generate podman containers I can then generate systemd units for, such that I can commit all the yaml for my running containers to a GitHub repo and easily recreate the state of the systemd units for whatever new version of podman might exist with whatever enhancements might exist for the `generate systemd` command. 

I'm trying to find what the proper way would be to support a use case where to edit my services I could tweak a bunch of yaml, run docker compose up, and then regenerate the systemd units, as opposed to having to run a bunch of individual `podman create` and `podman start` commands in bash scripts, and then run generate systemd from those running containers.

**Reply by vrothberg** [reply]
@alexlarsson started experimenting with https://github.com/containers/quadlet/ which I think would address that use case.

**Reply by vrothberg** [reply]
@rhatdan FYI

**Reply by andrew-kennedy** [reply]
Quadlet looks nice, but every container on the planet often includes docker-compose examples in the readme and having to use yet another syntax is kind of suboptimal. Though it's definitely better than nothing. It seems there really aren't many options for nice integrations with systemd, maybe I just need to move to a single node kubernetes setup...

**Comment by rhatdan** [comment]
The problem with docker-compose is that it is client service, and systemd prefers fork/exec.
As soon as you use something like docker-compose of docker within a systemd unit file, systemd looses it's view of what is going on.

**Reply by andrew-kennedy** [reply]
Hmm I guess quadlet is indeed the best option for a close-ish to docker compose type of solution. Hopefully it remains actively developed!

It's too bad there's nothing quite at the level of docker compose for podman as for a single node system it's quite unparalleled for quickly setting up services to always be up on restart and such.
